<h1>Содержание <span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Обзор данных</a></span></li><li><span><a href="#Заполнение-пропусков" data-toc-modified-id="Заполнение-пропусков-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Заполнение пропусков</a></span></li><li><span><a href="#Проверка-данных-на-аномалии-и-исправления." data-toc-modified-id="Проверка-данных-на-аномалии-и-исправления.-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Проверка данных на аномалии и исправления.</a></span></li><li><span><a href="#Изменение-типов-данных." data-toc-modified-id="Изменение-типов-данных.-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Изменение типов данных.</a></span></li><li><span><a href="#Удаление-дубликатов." data-toc-modified-id="Удаление-дубликатов.-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Удаление дубликатов.</a></span></li><li><span><a href="#Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма." data-toc-modified-id="Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма.-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.</a></span></li><li><span><a href="#Категоризация-дохода." data-toc-modified-id="Категоризация-дохода.-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Категоризация дохода.</a></span></li><li><span><a href="#Категоризация-целей-кредита." data-toc-modified-id="Категоризация-целей-кредита.-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Категоризация целей кредита.</a></span></li><li><span><a href="#Ответы-на-вопросы." data-toc-modified-id="Ответы-на-вопросы.-9"><span class="toc-item-num">9&nbsp;&nbsp;</span>Ответы на вопросы.</a></span><ul class="toc-item"><li><span><a href="#Вопрос-1:-Есть-ли-зависимость-между-наличием-детей-и-возвратом-кредита-в-срок?" data-toc-modified-id="Вопрос-1:-Есть-ли-зависимость-между-наличием-детей-и-возвратом-кредита-в-срок?-9.1"><span class="toc-item-num">9.1&nbsp;&nbsp;</span>Вопрос 1: Есть ли зависимость между наличием детей и возвратом кредита в срок?</a></span><ul class="toc-item"><li><span><a href="#Вывод-1:-Да,-зависимость-есть,-и-она-почти-прямо-пропорциональная.-В-семьях-без-детей-самая-низкая-вероятность-стать-должником-и-самая-высокая-в-семьях-с-четырьмя-детьми.-Но-почему-то-в-семьях-с-тремя-детьми-вероятность-ниже,-чем-в-семьях-с-одним-ребенком." data-toc-modified-id="Вывод-1:-Да,-зависимость-есть,-и-она-почти-прямо-пропорциональная.-В-семьях-без-детей-самая-низкая-вероятность-стать-должником-и-самая-высокая-в-семьях-с-четырьмя-детьми.-Но-почему-то-в-семьях-с-тремя-детьми-вероятность-ниже,-чем-в-семьях-с-одним-ребенком.-9.1.1"><span class="toc-item-num">9.1.1&nbsp;&nbsp;</span>Вывод 1: Да, зависимость есть, и она почти прямо пропорциональная. В семьях без детей самая низкая вероятность стать должником и самая высокая в семьях с четырьмя детьми. Но почему-то в семьях с тремя детьми вероятность ниже, чем в семьях с одним ребенком.</a></span></li></ul></li><li><span><a href="#Вопрос-2:-Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?" data-toc-modified-id="Вопрос-2:-Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?-9.2"><span class="toc-item-num">9.2&nbsp;&nbsp;</span>Вопрос 2: Есть ли зависимость между семейным положением и возвратом кредита в срок?</a></span><ul class="toc-item"><li><span><a href="#Вывод-2:-Да,-зависимость-есть.-Самая-низкая-вероятность-стать-должником-у-категории-вдовец-/-вдова-и-самая-высокая-у-категории-не-женат-/-не-замужем." data-toc-modified-id="Вывод-2:-Да,-зависимость-есть.-Самая-низкая-вероятность-стать-должником-у-категории-вдовец-/-вдова-и-самая-высокая-у-категории-не-женат-/-не-замужем.-9.2.1"><span class="toc-item-num">9.2.1&nbsp;&nbsp;</span>Вывод 2: Да, зависимость есть. Самая низкая вероятность стать должником у категории вдовец / вдова и самая высокая у категории не женат / не замужем.</a></span></li></ul></li><li><span><a href="#Вопрос-3:-Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?" data-toc-modified-id="Вопрос-3:-Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?-9.3"><span class="toc-item-num">9.3&nbsp;&nbsp;</span>Вопрос 3: Есть ли зависимость между уровнем дохода и возвратом кредита в срок?</a></span><ul class="toc-item"><li><span><a href="#Вывод-3:-Да,-зависимость-есть.-Самая-низкая-вероятность-стать-должником-у-заёмщиков-с-доходом-D-30001–50000-и-самая-высокая-у-заёмщиков-с-доходом-E-0–30000." data-toc-modified-id="Вывод-3:-Да,-зависимость-есть.-Самая-низкая-вероятность-стать-должником-у-заёмщиков-с-доходом-D-30001–50000-и-самая-высокая-у-заёмщиков-с-доходом-E-0–30000.-9.3.1"><span class="toc-item-num">9.3.1&nbsp;&nbsp;</span>Вывод 3: Да, зависимость есть. Самая низкая вероятность стать должником у заёмщиков с доходом D 30001–50000 и самая высокая у заёмщиков с доходом E 0–30000.</a></span></li></ul></li><li><span><a href="#Вопрос-4:-Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?" data-toc-modified-id="Вопрос-4:-Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?-9.4"><span class="toc-item-num">9.4&nbsp;&nbsp;</span>Вопрос 4: Как разные цели кредита влияют на его возврат в срок?</a></span><ul class="toc-item"><li><span><a href="#Вывод-4:-Да,-зависимость-есть.-Самая-низкая-вероятность-стать-должником-у-заёмщиков-с-целью-кредита-на-операции-с-недвижимостью--и-самая-высокая-с-целью-кредита-на-операции-с-автомобилем." data-toc-modified-id="Вывод-4:-Да,-зависимость-есть.-Самая-низкая-вероятность-стать-должником-у-заёмщиков-с-целью-кредита-на-операции-с-недвижимостью--и-самая-высокая-с-целью-кредита-на-операции-с-автомобилем.-9.4.1"><span class="toc-item-num">9.4.1&nbsp;&nbsp;</span>Вывод 4: Да, зависимость есть. Самая низкая вероятность стать должником у заёмщиков с целью кредита на операции с недвижимостью  и самая высокая с целью кредита на операции с автомобилем.</a></span></li></ul></li></ul></li><li><span><a href="#Общий-вывод:" data-toc-modified-id="Общий-вывод:-10"><span class="toc-item-num">10&nbsp;&nbsp;</span>Общий вывод:</a></span></li></ul></div>

# Тема: Предобработка данных
# Проект: Исследование надежности заемщиков.
# Описание проекта

* Заказчик — кредитный отдел банка.
* Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок.
* Входные данные от банка — статистика о платёжеспособности клиентов.
* Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

В ходе выполнения проекта нужно ответить на **4 вопроса**: 
* Есть ли зависимость между наличием детей и возвратом кредита в срок?
* Есть ли зависимость между семейным положением и возвратом кредита в срок?
* Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
* Как разные цели кредита влияют на его возврат в срок?

## Обзор данных

Импортируем библиотеку `pandas`.

In [1]:
import pandas as pd

Прочтем файл `data.csv` из папки `/datasets` и сохраните его в переменной `df`:

In [2]:
try:    
    df = pd.read_csv('D:/notebook/project_2/data.csv')   
except:
    df = pd.read_csv('/datasets/data.csv')

Выведем на экран первые пятнадцать строк таблицы:

In [3]:
df.head(15)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   

     total_income                            purpose  
0   253875.639453                      покупка жилья  
1   112080.014102            приобретение автомобиля  
2   145885.952297                      покупка жилья  
3   267628.550329         дополнительное образование  
4   158616.077870                    сыграть свадьбу  
5   255763.565419                      покупка жилья  
6   240525.971920                  операции с жильем  
7   135823.934197                        образование  
8    95856.832424              на проведение свадьбы  
9   144425.938277            покупка жилья для семьи  
10  113943.491460               покупка недвижимости  
11   77069.234271  покупка коммерческой недвижимости  
12            NaN                    сыграть свадьбу  
13  130458.228857            приобретение автомобиля  
14  165127.911772         покупка жилой недвижимости

Согласно документации к данным:
* `children` — количество детей в семье;
* `days_employed` — общий трудовой стаж в днях;
* `dob_years` — возраст клиента в годах;
* `education` — уровень образования клиента;
* `education_id` — идентификатор уровня образования;
* `family_status` — семейное положение;
* `family_status_id` — идентификатор семейного положения;
* `gender` — пол клиента;
* `income_type` — тип занятости;
* `debt` — имел ли задолженность по возврату кредитов;
* `total_income` — ежемесячный доход;
* `purpose` — цель получения кредита;

Получим общую информацию о таблице:

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Итак, в таблице двенадцать столбцов, три типа данных: `int64`, `float64`, `object`. Количество значений в столбцах различается. Значит, в данных есть пропущенные значения.

Посмотрим основные статистические характеристики данных по каждому числовому признаку.

In [5]:
df.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

В столбцах `children`, `days_employed`, `dob_years` имеются аномальные значения.

**Выводы**

В каждой строке таблицы — данные о потенциальном заёмщике и цель получения кредита. 

Предварительно можно утверждать, что, данных достаточно для проверки гипотез. Встречаются пропуски в данных и аномальные значения. Значения в столбце `education` используют разный регист.

Чтобы двигаться дальше, нужно устранить проблемы в данных.

## Заполнение пропусков

Проверим количество пропущенных значений в столбцах таблицы при помощи последовательного использования методов `isna()` и `sum()`.

In [6]:
df.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

Подозрительное совпадение количества пропусков в столбцах `days_employed` и `total_income`, закрадывается ощущение, что пропуски в данных столбцах находятся в одних и тех же строках, проверим это.

In [7]:
nan_employed_income = df.loc[(df['days_employed'].isna()==True) & (df['total_income'].isna()==True)]
# Создадим переменную, в которой будут храниться строчки с пропусками в обоих столбцах.
nan_employed_income_count = nan_employed_income['income_type'].count()
# Создадим переменную с количеством таких строк
display(nan_employed_income_count)

2174

Гипотеза подтвердилась.

Вычислим долю пропущенных значений в столбцах `days_employed` и `total_income`, результат вычислений сохраним в переменной `share_nan_employed_income`.

In [8]:
df.isna().mean()

children            0.000000
days_employed       0.100999
dob_years           0.000000
education           0.000000
education_id        0.000000
family_status       0.000000
family_status_id    0.000000
gender              0.000000
income_type         0.000000
debt                0.000000
total_income        0.100999
purpose             0.000000
dtype: float64

Доля пропущенных значений в столбцах `days_employed` и `total_income` составляет более 10 %.

Посмотрим, как распределены пропуски в зависимости от типа занятости заёмщиков, то есть от значений столбца `income_type`.

In [9]:
nan_employed_income.groupby('income_type')['income_type'].count().sort_values(ascending=False)

income_type
сотрудник          1105
компаньон           508
пенсионер           413
госслужащий         147
предприниматель       1
Name: income_type, dtype: int64

Посмотрим общее количество заёмщиков по типу занятости.

In [10]:
df['income_type'].value_counts()

сотрудник          11119
компаньон           5085
пенсионер           3856
госслужащий         1459
безработный            2
предприниматель        2
студент                1
в декрете              1
Name: income_type, dtype: int64

Пропуски распределены практически равномерно по типам занятости, в каждом типе, который насчитывает более 1400 потенциальных заёмщиков, примерно по 10% пропусков. Для групп "безработный", "предприниматель", "студент", "в декрете" доля пропусков не так важна ввиду, того, что суммарное количество заёмщиков в этих группах ничтожно мало относительно всего датафрейма, 6 человек из 21-ой тысячи человек.

При таком распределении пропуски лучше заполнить медианой, т.к. максимальные значения стажа и дохода у некоторых заёмщиков могут быть в разы выше средних, и в случае заполнения пропусков средним, значения данных показателей у заемщиков низкого и среднего сегмента окажутся сильно завышены.

Заполним пропуски в столбце `total_income`.

In [11]:
df['total_income'] = df['total_income'].fillna(df.groupby('income_type')['total_income'].transform("median"))

Заполним пропуски в столбце `days_employed` по такому же принципу.

In [12]:
df['days_employed'] = df['days_employed'].fillna(df.groupby('income_type')['days_employed'].transform("median"))

Проверим наличие пропусков. Получим общую информацию о таблице:

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     21525 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Пропуски отсутствуют.

In [14]:
df.head(15)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0  365213.306266         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   

     total_income                            purpose  
0   253875.639453                      покупка жилья  
1   112080.014102            приобретение автомобиля  
2   145885.952297                      покупка жилья  
3   267628.550329         дополнительное образование  
4   158616.077870                    сыграть свадьбу  
5   255763.565419                      покупка жилья  
6   240525.971920                  операции с жильем  
7   135823.934197                        образование  
8    95856.832424              на проведение свадьбы  
9   144425.938277            покупка жилья для семьи  
10  113943.491460               покупка недвижимости  
11   77069.234271  покупка коммерческой недвижимости  
12  118514.486412                    сыграть свадьбу  
13  130458.228857            приобретение автомобиля  
14  165127.911772         покупка жилой недвижимости

В 13-ой строчке (порядковый номер 12) раньше был потенциальный заёмщик с пропусками, теперь значение заменено на медиану.

**Вывод**

Ввиду равномерного распределения пропусков по типам занятости (примерно 10%) и общего количества пропусков 10%, скорее всего 10 % данных были добавлены к датафрему из другого источника, в котором отсутствовали столбцы "общий трудовой стаж в днях" и "ежемесячный доход".


## Проверка данных на аномалии и исправления.

Посмотрим основные статистические характеристики данных по каждому числовому признаку.

In [15]:
df.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   21525.000000  21525.000000  21525.000000   
mean       0.538908   63550.497071     43.293380      0.817236   
std        1.381587  141150.420058     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2570.047544     33.000000      1.000000   
50%        0.000000   -1355.683356     42.000000      1.000000   
75%        1.000000    -316.240646     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  2.152500e+04  
mean           0.972544      0.080883  1.652253e+05  
std            1.420324      0.272661  9.804367e+04  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.077982e+05  
50%            0.000000      0.000000  1.425944e+05  
75%            1.000000      0.000000  1.955499e+05  
max            4.000000      1.000000  2.265604e+06

Обработаем столбец `children`.

In [16]:
df['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

Значения 20 и -1 скорее всего технические ошибки, доля таких значений мала относительно всего датафрейма. Заменим 20 на 2, а -1 на 1.

In [17]:
df['children'].replace(20,2,inplace = True)
df['children'].replace(-1,1,inplace = True)
df['children'].value_counts()

0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

Аномальное количество детей устранено.

Изучим, столбец `days_employed` подробнее.

В столбце имеются адекватные значения по величине, но отрицательные, а также неадекватные по величине, но положительные, например 401755.400475 день это 1100 лет.

Посмотрим в каких типах занятости отрицательные и положительные значения стажа.

In [18]:
df[df['days_employed'] > 0] ['income_type'].value_counts()

пенсионер      3856
безработный       2
Name: income_type, dtype: int64

In [19]:
df[df['days_employed'] < 0] ['income_type'].value_counts()

сотрудник          11119
компаньон           5085
госслужащий         1459
предприниматель        2
студент                1
в декрете              1
Name: income_type, dtype: int64

Изучим стаж группы пенсионеры детальнее.

In [20]:
df.loc[(df['income_type'] == 'пенсионер')]['days_employed'].describe()

count      3856.000000
mean     365025.963652
std       19909.115430
min      328728.720605
25%      348802.360364
50%      365213.306266
75%      380751.301140
max      401755.400475
Name: days_employed, dtype: float64

Изучим стаж группы безработные детальнее.

In [21]:
df.loc[(df['income_type'] == 'безработный')]['days_employed']

3133     337524.466835
14798    395302.838654
Name: days_employed, dtype: float64

Порядок значений в группах пенсионеры и безработные ('days_employed' > 0) одинаковый (шестизначные значения).
Скорее всего эти значения получились из-за того, что стаж указан в часах, а не в днях.
Соответственно к данным группам можем применить одинаковое решения по обработке: разделим на 24 часа. 

In [22]:
df.loc[(df['days_employed'] > 0), 'days_employed'] = (df['days_employed'])/24
df.days_employed.describe()

count    21525.000000
mean       851.546834
std       7007.842448
min     -18388.949901
25%      -2570.047544
50%      -1355.683356
75%       -316.240646
max      16739.808353
Name: days_employed, dtype: float64

Избавимся от отрицательных значений дней стажа в столбце `days_employed`.

In [23]:
df.loc[(df['days_employed'] < 0), 'days_employed'] = df['days_employed'].abs() 
df.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   21525.000000  21525.000000  21525.000000   
mean       0.479721    4600.535795     43.293380      0.817236   
std        0.755528    5354.354167     12.574584      0.548138   
min        0.000000      24.141633      0.000000      0.000000   
25%        0.000000    1025.549623     33.000000      1.000000   
50%        0.000000    1993.522017     42.000000      1.000000   
75%        1.000000    5347.024506     53.000000      1.000000   
max        5.000000   18388.949901     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  2.152500e+04  
mean           0.972544      0.080883  1.652253e+05  
std            1.420324      0.272661  9.804367e+04  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.077982e+05  
50%            0.000000      0.000000  1.425944e+05  
75%            1.000000      0.000000  1.955499e+05  
max            4.000000      1.000000  2.265604e+06

Отрицательных и аномальных значений больше нет.

**Вывод:** 

В таблице присутствовали отрицательные значения и технические ошибки, которые мы устранили.

## Изменение типов данных.

Изменим тип данных для столбца `total_income`

In [24]:
df['total_income'] = df['total_income'].astype('int64')
df.info()
display(df.head(5))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     21525 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  int64  
 11  purpose           21525 non-null  object 
dtypes: float64(1), int64(6), object(5)
memory usage: 2.0+ MB


children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   Среднее             1   
3         3    4124.747207         32   среднее             1   
4         0   14177.753002         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

Тип данных изменен.

## Удаление дубликатов.

Посмотрим, как записаны значения в столбце `education`.

In [25]:
df['education'].unique()

array(['высшее', 'среднее', 'Среднее', 'СРЕДНЕЕ', 'ВЫСШЕЕ',
       'неоконченное высшее', 'начальное', 'Высшее',
       'НЕОКОНЧЕННОЕ ВЫСШЕЕ', 'Неоконченное высшее', 'НАЧАЛЬНОЕ',
       'Начальное', 'Ученая степень', 'УЧЕНАЯ СТЕПЕНЬ', 'ученая степень'],
      dtype=object)

Приведем значения к общему регистру.

In [26]:
df['education'] = df['education'].str.lower()
df['education'].unique()

array(['высшее', 'среднее', 'неоконченное высшее', 'начальное',
       'ученая степень'], dtype=object)

Избавились от большей части дубликатов в данных.

Отформатируем нужные нам столбцы таким же методом.

In [27]:
df['family_status'].unique()

array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'Не женат / не замужем'], dtype=object)

In [28]:
df['family_status'] = df['family_status'].str.lower()
df['family_status'].unique()

array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'не женат / не замужем'], dtype=object)

In [29]:
df['income_type'].unique()

array(['сотрудник', 'пенсионер', 'компаньон', 'госслужащий',
       'безработный', 'предприниматель', 'студент', 'в декрете'],
      dtype=object)

В данном столбце регистр везде одинаковый.

Выясним количество дубликатов.

In [30]:
df.duplicated().sum()

71

Удалим дубликаты и назначим новую нумерацию.

In [31]:
df = df.drop_duplicates().reset_index(drop = True)
df.duplicated().sum()

0

**Вывод:** 
1. Исправлен регистр в значениях интересующих нас столбцов, теперь подсчеты по данным столбцам станут точнее, для устранения таких дубликатов использовался метод `unique`.
2. Удалены явные дубликаты, для их обнаружения использовалось последовательное применение методов `duplicated` и `sum`, а для удаления `drop_duplicates`, нумерация была задана новая функцией `reset_index`.

## Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.

Создадим новый датафрейм для уровня образования и его идентификатора.

In [32]:
df_education = df[['education','education_id']]
display(df_education.value_counts()) # Посмотрим какое количество заёмщиков соответствуют каждому уровню образования

education            education_id
среднее              1               15172
высшее               0                5250
неоконченное высшее  2                 744
начальное            3                 282
ученая степень       4                   6
dtype: int64

Создадим новый датафрейм для семейного статуса и его идентификатора.

In [33]:
df_family_status = df[['family_status','family_status_id']]
display(df_family_status.value_counts()) # Посмотрим какое количество заёмщиков соответствуют каждому семейному статусу

family_status          family_status_id
женат / замужем        0                   12339
гражданский брак       1                    4151
не женат / не замужем  4                    2810
в разводе              3                    1195
вдовец / вдова         2                     959
dtype: int64

Удалим из исходного датафрейма столбцы `education` и `family_status`, оставив только их идентификаторы: `education_id` и `family_status_id`.

In [34]:
df = df.drop(['education', 'family_status'], axis=1)

In [35]:
df.info()
display(df)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21454 entries, 0 to 21453
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21454 non-null  int64  
 1   days_employed     21454 non-null  float64
 2   dob_years         21454 non-null  int64  
 3   education_id      21454 non-null  int64  
 4   family_status_id  21454 non-null  int64  
 5   gender            21454 non-null  object 
 6   income_type       21454 non-null  object 
 7   debt              21454 non-null  int64  
 8   total_income      21454 non-null  int64  
 9   purpose           21454 non-null  object 
dtypes: float64(1), int64(6), object(3)
memory usage: 1.6+ MB


children  days_employed  dob_years  education_id  family_status_id  \
0             1    8437.673028         42             0                 0   
1             1    4024.803754         36             1                 0   
2             0    5623.422610         33             1                 0   
3             3    4124.747207         32             1                 0   
4             0   14177.753002         53             1                 1   
...         ...            ...        ...           ...               ...   
21449         1    4529.316663         43             1                 1   
21450         0   14330.725172         67             1                 0   
21451         1    2113.346888         38             1                 1   
21452         3    3112.481705         38             1                 0   
21453         2    1984.507589         40             1                 0   

      gender income_type  debt  total_income                       purpose  
0          F   сотрудник     0        253875                 покупка жилья  
1          F   сотрудник     0        112080       приобретение автомобиля  
2          M   сотрудник     0        145885                 покупка жилья  
3          M   сотрудник     0        267628    дополнительное образование  
4          F   пенсионер     0        158616               сыграть свадьбу  
...      ...         ...   ...           ...                           ...  
21449      F   компаньон     0        224791             операции с жильем  
21450      F   пенсионер     0        155999          сделка с автомобилем  
21451      M   сотрудник     1         89672                  недвижимость  
21452      M   сотрудник     1        244093  на покупку своего автомобиля  
21453      F   сотрудник     0         82047         на покупку автомобиля  

[21454 rows x 10 columns]

Разгрузили таблицу от лишней информации.

## Категоризация дохода.

In [36]:
def total_income_category (income):
    """
    Возвращает группу по значению дохода total_income, используя правила:
    0–30000 — 'E';
    30001–50000 — 'D';
    50001–200000 — 'C';
    200001–1000000 — 'B';
    1000001 и выше — 'A'.
    """
    if income <= 30000:
        return 'E'
    if income <= 50000:
        return 'D'
    if income <= 200000:
        return 'C'
    if income <= 1000000:
        return 'B'
    return 'A'
df['total_income_category'] = df['total_income'].apply(total_income_category)
display(df.head())

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1    8437.673028         42             0                 0      F   
1         1    4024.803754         36             1                 0      F   
2         0    5623.422610         33             1                 0      M   
3         3    4124.747207         32             1                 0      M   
4         0   14177.753002         53             1                 1      F   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   

  total_income_category  
0                     B  
1                     C  
2                     C  
3                     B  
4                     C

## Категоризация целей кредита.

In [37]:
def purpose_category (purpose):
    if 'автомобил' in purpose:
        return 'операции с автомобилем'
    if 'недвижимост' in purpose or 'жиль' in purpose:
        return 'операции с недвижимостью'
    if 'свадьб' in purpose:
        return 'проведение свадьбы'
    if 'образовани' in purpose:
        return 'получение образования'    
df['purpose_category'] = df['purpose'].apply(purpose_category)
display(df.head(50))        
        

children  days_employed  dob_years  education_id  family_status_id gender  \
0          1    8437.673028         42             0                 0      F   
1          1    4024.803754         36             1                 0      F   
2          0    5623.422610         33             1                 0      M   
3          3    4124.747207         32             1                 0      M   
4          0   14177.753002         53             1                 1      F   
5          0     926.185831         27             0                 1      M   
6          0    2879.202052         43             0                 0      F   
7          0     152.779569         50             1                 0      M   
8          2    6929.865299         35             0                 1      F   
9          0    2188.756445         41             1                 0      M   
10         2    4171.483647         36             0                 0      M   
11         0     792.701887         40             1                 0      F   
12         0   15217.221094         65             1                 1      M   
13         0    1846.641941         54             2                 0      F   
14         0    1844.956182         56             0                 1      F   
15         1     972.364419         26             1                 0      F   
16         0    1719.934226         35             1                 0      F   
17         0    2369.999720         33             0                 1      M   
18         0   16678.380705         53             1                 2      F   
19         0   10038.818549         48             1                 3      F   
20         1    1311.604166         36             1                 0      M   
21         1     253.685166         33             1                 1      F   
22         1    1766.644138         24             1                 1      F   
23         0     272.981385         21             0                 1      M   
24         1   14106.331371         57             1                 4      F   
25         0   15147.853723         67             1                 0      M   
26         0    2689.368353         41             1                 0      M   
27         0     529.191635         28             0                 0      M   
28         1     717.274324         26             0                 0      F   
29         0   15217.221094         63             1                 4      F   
30         1   13982.569521         62             1                 0      F   
31         0    1682.083438         47             3                 0      F   
32         0    4649.910832         34             1                 1      F   
33         0    1548.637544         48             1                 0      F   
34         0    4488.067031         35             1                 0      F   
35         0   16417.544674         68             1                 1      M   
36         0     176.216688         33             1                 4      M   
37         0    6448.810860         43             0                 1      F   
38         0     597.881827         25             0                 4      M   
39         0     650.587796         31             0                 1      F   
40         0    1030.362606         30             1                 4      M   
41         0    2689.368353         50             1                 0      F   
42         0    1257.496190         20             2                 4      F   
43         0    4375.681384         43             2                 3      F   
44         1    1362.041728         26             1                 0      F   
45         1    1039.451741         49             1                 4      F   
46         2    2262.712304         37             1                 0      F   
47         1    2689.137274         33             0                 1      M   
48         0    3341.067886        

## Ответы на вопросы.

### Вопрос 1: Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [38]:
pivot_table_one = df.pivot_table(index=['children'],
                                values = 'debt',
                                aggfunc='mean').sort_values(by='debt')
display(pivot_table_one)

debt
children          
5         0.000000
0         0.075438
3         0.081818
1         0.091658
2         0.094925
4         0.097561

#### Вывод 1: Да, зависимость есть, и она почти прямо пропорциональная. В семьях без детей самая низкая вероятность стать должником и самая высокая в семьях с четырьмя детьми. Но почему-то в семьях с тремя детьми вероятность ниже, чем в семьях с одним ребенком.

###  Вопрос 2: Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [39]:
pivot_table_two = df.pivot_table(index=['family_status_id'],
                                values = 'debt',
                                aggfunc='mean').sort_values(by='debt')
display(pivot_table_two)

debt
family_status_id          
2                 0.065693
3                 0.071130
0                 0.075452
1                 0.093471
4                 0.097509

In [40]:
df_family_status.value_counts()

family_status          family_status_id
женат / замужем        0                   12339
гражданский брак       1                    4151
не женат / не замужем  4                    2810
в разводе              3                    1195
вдовец / вдова         2                     959
dtype: int64

#### Вывод 2: Да, зависимость есть. Самая низкая вероятность стать должником у категории вдовец / вдова и самая высокая у категории не женат / не замужем.

### Вопрос 3: Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [41]:
pivot_table_three = df.pivot_table(index=['total_income_category'],
                                values = 'debt',
                                aggfunc='mean').sort_values(by='debt')
display(pivot_table_three)

debt
total_income_category          
D                      0.060000
B                      0.070607
A                      0.080000
C                      0.084920
E                      0.090909

In [42]:
df['total_income_category'].value_counts()

C    16015
B     5042
D      350
A       25
E       22
Name: total_income_category, dtype: int64

1. A 1000001 и выше
2. B 200001–1000000
3. C 50001–200000
4. D 30001–50000
5. E 0–30000

#### Вывод 3: Да, зависимость есть. Самая низкая вероятность стать должником у заёмщиков с доходом D 30001–50000 и самая высокая у заёмщиков с доходом E 0–30000.

### Вопрос 4: Как разные цели кредита влияют на его возврат в срок?

In [43]:
pivot_table_four = df.pivot_table(index=['purpose_category'],
                                values = 'debt',
                                aggfunc='mean').sort_values(by='debt')
display(pivot_table_four)

debt
purpose_category                  
операции с недвижимостью  0.072334
проведение свадьбы        0.080034
получение образования     0.092200
операции с автомобилем    0.093590

#### Вывод 4: Да, зависимость есть. Самая низкая вероятность стать должником у заёмщиков с целью кредита на операции с недвижимостью  и самая высокая с целью кредита на операции с автомобилем.

## Общий вывод:
Семейное положение и наличие детей в семье сильно влияют на погашение кредита. Для банка наибольший риск просрочки оплаты по кредиту при выдаче кредита клиентам имеющим 4х детей, не женатых/не замужем, с доходом от 0 до 30000, с целью кредита на операции с автомобилем, наименьший риск для банка при выдаче кредита клиентам без детей, вдовцам/вдовам, с доходом от 30000-50000, с целью кредита на операции с недвижимостью.